app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt

```diff
@@ -0,0 +1,168 @@
+  private lateinit var value: String
+  private lateinit var units: String
+  private lateinit var CURRENCY_UNITS: HashMap<String, Units>
+
+  /** @return index of pattern in s or -1, if not found
+   */
+  fun indexOf(pattern: Pattern, s: String): Int {
+    val matcher = pattern.matcher(s)
+    return if (matcher.find()) matcher.start() else -1
+  }
+
+  fun getCurrencyUnits(): CurrencyUnit {
+    var currencyUnits = HashMap<String, Units>()
+    currencyUnits.put(
+      "dollar",
+      Units("dollar", mutableListOf("$", "dollars", "Dollars", "Dollar", "USD"), mutableListOf("$"), null)
+    )
+    currencyUnits.put(
+      "rupee",
+      Units("rupee", mutableListOf("Rs", "rupees", "₹", "Rupees", "Rupee"), mutableListOf("Rs ", "₹"), null)
+    )
+    currencyUnits.put(
+      "cent", Units("cent", mutableListOf("cents", "Cents", "Cent"), mutableListOf(), "0.01 dollar")
+    )
+    currencyUnits.put(
+      "paise", Units("paise", mutableListOf("paisa", "Paise", "Paisa"), mutableListOf(), "0.01 rupee")
+    )
+    return CurrencyUnit(currencyUnits)
+  }
+
+  fun getNumberWithUnits(inputText: String): NumberWithUnits {
+    var rawInput: String = inputText.normalizeWhitespace()
+    CURRENCY_UNITS = getCurrencyUnits().unitHashMap
+    rawInput = rawInput.trim()
+    var type = ""
+    var real = 0.0
+    // Default fraction value.
+    var fractionObj = Fraction.newBuilder().setDenominator(1).build()
+    var units = ""
+    var value = ""
+
+    // Allow validation only when rawInput is not null or an empty string.
+    if (rawInput.isNotEmpty() && rawInput != null) {
+      // Start with digit when there is no currency unit.
+      if (rawInput.matches("^\\-?\\d.*\$".toRegex())) {
+        var ind = indexOf(Pattern.compile("[a-z(₹$]"), rawInput)
+        if (ind == -1) {
+          // There is value with no units.
+          value = rawInput;
+          units = ""
+        } else {
+          ind -= 1
+          value = rawInput.substring(0, ind).trim()
+          units = rawInput.substring(ind).trim()
+        }
+
+        var keys = (CURRENCY_UNITS).keys
+        for (i in keys) {
+          for (j in 0 until CURRENCY_UNITS[i]!!.frontUnits.size) {
+            if (units.indexOf(CURRENCY_UNITS[i]!!.frontUnits[j]) != -1) {
+              throw  Error(
+                "INVALID_CURRENCY_FORMAT"
+              )
+            }
+          }
+        }
+      } else {
+        var startsWithCorrectCurrencyUnit = false
+        var keys = (CURRENCY_UNITS).keys
+        for (i in keys) {
+          for (j in 0 until CURRENCY_UNITS[i]!!.frontUnits.size) {
+            if (rawInput.startsWith(CURRENCY_UNITS[i]!!.frontUnits[j])) {
+              startsWithCorrectCurrencyUnit = true
+              break
+            }
+          }
+        }
+        if (startsWithCorrectCurrencyUnit == false) {
+          throw  Error("INVALID_CURRENCY")
+        }
+
+        var ind = indexOf(Pattern.compile("[0-9]"), rawInput)
+
+
+        if (ind === -1) {
```

Comment:
We should be using ``match`` methods to check presence & extract values rather than doing index-based parsing, similar to what we do in fractions (e.g. https://github.com/oppia/oppia-android/blob/develop/app/src/main/java/org/oppia/app/parser/StringToFractionParser.kt#L83). I think this will simplify a lot of the logic here since I think some of these things can be replaced with patterns (e.g. checking that the value starts with the correct unit).

Reply:
As per your instructions in meeting we are using this to get started.